feat: serve workspacekind assets (logos + icons)

andyatmiami opened the pull request:
closes: #692 

This commit implements serving of workspacekind assets (logos and icons) from ConfigMaps through the backend API, with SHA256 hash support for cache-busting.

Key changes:
- Add endpoint to serve workspacekind assets from ConfigMaps at `/api/v1/workspacekinds/{name}/assets/{icon|logo}.svg`
- Compute and include SHA256 hashes for ConfigMap-based assets as query parameters for cache-busting
- Add ConfigMap client support to repositories for accessing asset content
    - partial cache for performance/efficiency that keys off label:
        - `notebooks.kubeflow.org/image-source: true`
- Add GetConfigMapContent method to `WorkspaceKindRepository`  and `WorkspaceRepository` for retrieving asset content from ConfigMaps
- Update model functions to support SHA256 hashes in asset URLs:
  - NewWorkspaceKindModelFromWorkspaceKindWithAssetHashes
  - NewWorkspaceModelFromWorkspaceWithAssetHashes
- Added workspacekind_imagesource_configmap.yaml for reference example
- Refactored `backend` `mgr` setup in `main.go` to leverage `helper/k8s.go`
    - easier to keep logic in sync in e2e tests - particularly with new partial cache

Remote URL specification of assets is still supported - but the browser queries
the source URL directly for content.


#### SCREENSHOTS
<img width="2065" height="390" alt="image" src="https://github.com/user-attachments/assets/d54e44fb-d6e1-4ace-a751-58c24188e82a" />
<img width="1992" height="262" alt="image" src="https://github.com/user-attachments/assets/176103ad-1869-441c-97fc-155e01f091a2" />
<img width="1995" height="592" alt="image" src="https://github.com/user-attachments/assets/603c3153-0ff3-4c1a-9d5a-8b12d1c6ab53" />

andyatmiami committed:
feat: serve workspacekind assets (logos + icons)

Add support for serving WorkspaceKind icon and logo assets from
Kubernetes ConfigMaps, with SHA256 content hashing for cache-busting.

Controller:
- Add ImageAssetStatus, WorkspaceKindConfigMapStatus, and ConfigMapError
  types to WorkspaceKind status for tracking asset state
- Add reconcileAssetStatus() to compute SHA256 hashes from ConfigMap
  content via a filtered in-memory cache, reporting errors for missing
  ConfigMaps or keys
- Build a label-filtered ConfigMap cache (notebooks.kubeflow.org/
  image-source=true) with a Transform function that pre-computes SHA256
  hashes as virtual annotations
- Watch ConfigMap changes to re-reconcile referencing WorkspaceKinds
- Add namespace field to WorkspaceKindConfigMap for cross-namespace
  ConfigMap references
- Add envtest integration tests for ConfigMap-based asset reconciliation
  covering hash population and missing ConfigMap error reporting

Backend:
- Add GET endpoint at /api/v1/workspacekinds/{name}/assets/{assetType}
  to serve asset content directly from ConfigMaps
- Add GetConfigMapContent to WorkspaceKindRepository using a shared
  partial ConfigMap cache for efficient lookups
- Include SHA256 query parameters in asset URLs for browser cache-busting
- Refactor asset URL construction into common/assets package with
  support for both URL-based and ConfigMap-based assets
- Refactor manager setup in main.go to use shared helper/k8s.go for
  cache construction, keeping backend and e2e test setup in sync
- Add unit tests for asset URL building and handler integration tests

Assisted-by: Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Andy Stoneberg <astonebe@redhat.com>

andyatmiami committed:
feat: serve workspacekind assets (logos + icons)

Add support for serving WorkspaceKind icon and logo assets from
Kubernetes ConfigMaps, with SHA256 content hashing for cache-busting.

Controller:
- Add ImageAssetStatus, WorkspaceKindConfigMapStatus, and ConfigMapError
  types to WorkspaceKind status for tracking asset state
- Add reconcileAssetStatus() to compute SHA256 hashes from ConfigMap
  content via a filtered in-memory cache, reporting errors for missing
  ConfigMaps or keys
- Build a label-filtered ConfigMap cache (notebooks.kubeflow.org/
  image-source=true) with a Transform function that pre-computes SHA256
  hashes as virtual annotations
- Watch ConfigMap changes to re-reconcile referencing WorkspaceKinds
- Add namespace field to WorkspaceKindConfigMap for cross-namespace
  ConfigMap references
- Add envtest integration tests for ConfigMap-based asset reconciliation
  covering hash population and missing ConfigMap error reporting

Backend:
- Add GET endpoint at /api/v1/workspacekinds/{name}/assets/{assetType}
  to serve asset content directly from ConfigMaps
- Add GetConfigMapContent to WorkspaceKindRepository using a shared
  partial ConfigMap cache for efficient lookups
- Include SHA256 query parameters in asset URLs for browser cache-busting
- Refactor asset URL construction into common/assets package with
  support for both URL-based and ConfigMap-based assets
- Refactor manager setup in main.go to use shared helper/k8s.go for
  cache construction, keeping backend and e2e test setup in sync
- Add unit tests for asset URL building and handler integration tests

Assisted-by: Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Andy Stoneberg <astonebe@redhat.com>

thaorell reviewed (changes requested):
While reviewing the implementation for `/api/v1/workspacekinds/{name}/assets/{logo|icon|.svg`, I did a quick scan of the existing data flow.
It looks like the frontend/client already receives the full source URL for these icons and logos from ImageRef as part of the parent `WorkspaceKind` object or a payload from `WorkspaceKindInfo` in loading workspace(s).

If that's the case, could we simply have the browser render those URLs directly? That would allow us to avoid maintaining a dedicated proxy endpoint and increase maintainability.

Let me know if I’m missing a specific use case where this proxy is required

christian-heusel reviewed (changes requested):
Tested with the changes from #906, it seems like we also need to account for potentially differently sized images in the frontend:
<img width="2255" height="1329" alt="image" src="https://github.com/user-attachments/assets/cf2a9a5e-34ca-4c19-b931-ec80ea708fc5" />

<img width="2255" height="1329" alt="Image" src="https://github.com/user-attachments/assets/824a7804-b539-41eb-8bb8-5d17a2b29896" />

thesuperzapper committed:
mathew: 1

Signed-off-by: Mathew Wicks <5735406+thesuperzapper@users.noreply.github.com>

thesuperzapper commented:
@andyatmiami after our review I have pushed https://github.com/kubeflow/notebooks/pull/736/commits/eed53427e6f83353e780e7eb06f3b08eb52f5621 with changes discussed. 

You should test and rebase so we can merge soon.

thesuperzapper committed:
mathew: 2

Signed-off-by: Mathew Wicks <5735406+thesuperzapper@users.noreply.github.com>

thesuperzapper commented:
@andyatmiami I rebased it and created https://github.com/kubeflow/notebooks/pull/736/commits/73863995c61f4e66a2816877597eb2fa0a125ea8 to bring the new code up to speed with the changes/refactor of constants.

The one strange thing is that at some point in the last few commits we have broken the swagger UI (at least in tilt), and we now get un-templated fields at the top, I think this happens at top of tree right now anyway (so probably should not block this PR), but we should figure out what happend and fix it.

<img width="675" height="305" alt="Screenshot 2026-05-01 at 01 38 51" src="https://github.com/user-attachments/assets/92709c23-e1b3-4434-b3a9-0dd8133bd991" />

thesuperzapper commented:
Even just before my own rebase it was fine when at the commit https://github.com/kubeflow/notebooks/pull/736/changes/eed53427e6f83353e780e7eb06f3b08eb52f5621
<img width="697" height="304" alt="Screenshot 2026-05-01 at 01 41 31" src="https://github.com/user-attachments/assets/da0a2b13-1c77-41cb-8c82-aafd89f13fb0" />

andyatmiami committed:
chore: minor fixes based on PR review commits

Address review feedback for the WorkspaceKind assets feature:

- Fix swag template delimiters to use '{{{{,}}}}' so Go template expressions in generated docs.go are not interpreted at runtime by the Swagger library's own template engine
- Add PROXY_URL_PREFIX env var to Tiltfile for backend development
- Separate path and query parameter validation in the asset handler, using distinct error messages for each category
- Remove blank import grouping line in asset handler imports
- Replace placeholder TODO with comprehensive test suite covering  ConfigMap-based and URL-based assets, auth, and validation

Assisted-by: Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Andy Stoneberg <astonebe@redhat.com>

thesuperzapper committed:
mathew: 3

Signed-off-by: Mathew Wicks <5735406+thesuperzapper@users.noreply.github.com>
